### PR TITLE
Merge-Tree and SharedString ISegment Deprecations

### PR DESCRIPTION
**Pull request by anthony-murphy**
The current ISegment interface over-exposes a number of properties which do not have an external use case, and any external usage could result in damage to the underlying merge-tree including data corruption.

The only use case that will continue to be supported is determining if a segment is removed. For this purpose we've add the following `function segmentIsRemoved(segment: ISegment): boolean`

For example, checking if a segment is not removed would change as follows:
``` diff
- if(segment.removedSeq === undefined){
+ if(!segmentIsRemoved(segment)){
```

The following properties are deprecated on ISegment and its implementations:
- clientId
- index
- localMovedSeq
- localRefs
- localRemovedSeq
- localSeq
- movedClientsIds
- movedSeq
- movedSeqs
- ordinal
- removedClientIds
- removedSeq
- seq
- wasMovedOnInsert

Additionally, the following types are also deprecated, and will become internal:
- IMergeNodeCommon
- IMoveInfo
- IRemovalInfo
- LocalReferenceCollection

### CODE REVIEW

File: .changeset/real-grapes-kick.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+---
+---
+"section": deprecation
+---
+
+Merge-Tree and SharedString ISegment Deprecations
+
+The current ISegment interface over-exposes a number of properties which do not have an external use case, and any external usage could result in damage to the underlying merge-tree including data corruption.
+
+The only use case that will continue to be supported is determining if a segment is removed. For this purpose we've add the following `function segmentIsRemoved(segment: ISegment): boolean`
```
Comment (alexvy86) [inline .changeset/real-grapes-kick.md:13]: Wording nit
```suggestion
The only use case that will continue to be supported is determining if a segment is removed. For this purpose we've added the free function `segmentIsRemoved(segment: ISegment): boolean`.
```

File: .changeset/real-grapes-kick.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+- movedClientsIds
+- movedSeq
+- movedSeqs
+- ordinal
+- removedClientIds
+- removedSeq
+- seq
+- wasMovedOnInsert
+
+Additionally, the following types are also deprecated, and will become internal:
```
Comment (alexvy86) [inline .changeset/real-grapes-kick.md:37]: Just for extra clarity on the implications for consumers, so people who aren't familiar with our API levels know what this means for them.
```suggestion
Additionally, the following types are also deprecated, and will become internal (i.e. users of the Fluid Framework will not have access to them):
```

**Review by alexvy86 — APPROVED** [review]
Aproving for docs with a couple of suggestions.

File: packages/dds/merge-tree/src/mergeTreeNodes.ts [hunk 3]
```diff
@@ -292,6 +307,73 @@ export interface ISegment extends IMergeNodeCommon, Partial<IRemovalInfo>, Parti
...
+ * Determine if a segment has been removed.
+ * @legacy
+ * @alpha
+ */
+export function segmentIsRemoved(segment: ISegment): boolean {
+	const leaf: ISegmentLeaf = segment;
+	if (leaf.removedSeq === undefined) {
+		return false;
+	}
+	return true;
```
Comment (Abe27342) [inline packages/dds/merge-tree/src/mergeTreeNodes.ts:376]:
```suggestion
	return leaf.removedSeg !== undefined;
```

**Review by Abe27342 — APPROVED** [review]

**Pull request merged**